eZ/Publish/Core/FieldType/Url/UrlStorage/Gateway/DoctrineStorage.php

```diff
@@ -129,6 +129,48 @@ public function insertUrl($url)
         );
     }
 
+    /**
+     * Return a list of URLs used by the given field and version.
+     *
+     * @return bool[] An array of URLs, with urls as keys
+     */
+    public function getUrlsFromUrlLink(int $fieldId, int $versionNo): array
+    {
+        $selectQuery = $this->connection->createQueryBuilder();
+        $selectQuery
+            ->select($this->connection->quoteIdentifier('url.url'))
+            ->from($this->connection->quoteIdentifier(self::URL_TABLE), 'url')
+            ->leftJoin(
+                'url',
+                $this->connection->quoteIdentifier(self::URL_LINK_TABLE),
+                'link',
+                'url.id = link.url_id'
+            )
+            ->where(
+                $selectQuery->expr()->eq(
+                    'link.contentobject_attribute_id',
+                    ':contentobject_attribute_id'
+                )
+            )
+            ->andWhere(
+                $selectQuery->expr()->eq(
+                    'link.contentobject_attribute_version',
+                    ':contentobject_attribute_version'
+                )
+            )
+            ->setParameter(':contentobject_attribute_id', $fieldId, ParameterType::INTEGER)
+            ->setParameter(':contentobject_attribute_version', $versionNo, ParameterType::INTEGER);
+
+        $statement = $selectQuery->execute();
+        $rows = $statement->fetchAllAssociativeIndexed();
+        $result = [];
+        foreach ($rows as $url => $item) {
```

Comment:
It could be an option to return the array as-is, and instead to this foreach in [`vendor/ezsystems/ezplatform-richtext/src/lib/eZ/FieldType/RichText/RichTextStorage/Gateway.php`](https://github.com/ezsystems/ezplatform-richtext/pull/242/files#diff-c54260f9f2ffb301f5dfffd2821f50f16cceca1d7b05fcd2c3fcca31c4e4dc51)

Reply:
> It could be an option to return the array as-is, and instead to this foreach in [`vendor/ezsystems/ezplatform-richtext/src/lib/eZ/FieldType/RichText/RichTextStorage/Gateway.php`](https://github.com/ezsystems/ezplatform-richtext/pull/242/files#diff-c54260f9f2ffb301f5dfffd2821f50f16cceca1d7b05fcd2c3fcca31c4e4dc51)

If it's possible without breaking changes, that's definitely preferred.

ATM to get current result you could do instead
```php
        $rows = $statement->fetchFirstColumn();
        $result = [];
        foreach ($rows as $url) {
            $result[$url] = true;
        }
```
to reduce time & memory complexity (`fetchAllAssociativeIndexed` iterates over a result and then you iterate over it again accumulating more memory and getting 2 * O(k) time instead of O(k)).

But as said, if it's possible to use simple return here and indeed you need this indexed by URL, so you can do `isset` later on, then maybe it's a better option.

---

Another idea: I see the code using this is as follows:
```php
$urlIdMap = $this->gateway->getUrlIdMap( // ...
// ...
$urlLinkSet = $this->gateway->getUrlsFromUrlLink( // ...
foreach ($links as $index => $link) {
// ..

                if (!isset($urlIdMap[$url])) {
                    $urlIdMap[$url] = $this->gateway->insertUrl($url);
                }
                if (!isset($urlLinkSet[$url])) {
                    $this->gateway->linkUrl(
                        $urlIdMap[$url],
                        $field->id,
                        $versionInfo->versionNo
                    );
                    $urlLinkSet[$url] = true;
                }
// ...
```
What if instead of calling `getUrlIdMap` and `getUrlsFromUrlLink` we could call one method which would give us a map of `url => id` and use that there in richtext?

Reply:
> What if instead of calling getUrlIdMap and getUrlsFromUrlLink we could call one method which would give us a map of url => id and use that there in richtext?

I find this tricky, sql wise

Consider the following row set:
```
MariaDB [ezp]> select * from ezurl as url left join ezurl_object_link as link on url.id = link.url_id where url.url in ("https://ibexa.co", "https://ez.no", "https://newurl");
+----+------------+----------+--------------+------------+----------------------------------+------------------+----------------------------+---------------------------------+--------+
| id | created    | is_valid | last_checked | modified   | original_url_md5                 | url              | contentobject_attribute_id | contentobject_attribute_version | url_id |
+----+------------+----------+--------------+------------+----------------------------------+------------------+----------------------------+---------------------------------+--------+
|  5 | 1705676561 |        1 |            0 | 1705676561 | f24389c80f5500a24427031fd45d283d | https://ibexa.co |                     183982 |                               1 |      5 |
|  5 | 1705676561 |        1 |            0 | 1705676561 | f24389c80f5500a24427031fd45d283d | https://ibexa.co |                     183702 |                               2 |      5 |
|  4 | 1705676559 |        1 |            0 | 1705676559 | d74110041197e107722d8821f5f4d89c | https://ez.no    |                     183703 |                               2 |      4 |
|  6 | 1705676559 |        1 |            0 | 1705676559 | aabbvv041197e107722d8821f5f4d89c | /only_in_url     |                       NULL |                            NULL |   NULL |
|  3 | 1705676559 |        1 |            0 | 1705676559 | f24389c80f5500a24427031fd45d283d | https://ibexa.co |                       NULL |                            NULL |   NULL |
+----+------------+----------+--------------+------------+----------------------------------+------------------+----------------------------+---------------------------------+--------+
```

Let's say you want to find the propriate rows for `contentobject_attribute_id=183702`
You'll need to adjust the query so that:
- you get the 2nd row (since contentobject_attribute_id matches 183702)
- you get the 3rd and 4th row (because `https://ez.no` do not have any other rows where contentobject_attribute_id=183702). Maybe possible using a advanced subquery and !EXISTS()?
Maybe you have an idea? if not, I can still persuit this approach if you insist even though I am not so sure it will have much benifit.

(btw, not sure why I have two rows in ezurl table with same url (`url=https://ibexa.co`, id=`3` and id=`5`) but this seems also possible for whatever reason

Reply:
> > What if instead of calling getUrlIdMap and getUrlsFromUrlLink we could call one method which would give us a map of url => id and use that there in richtext?
> 
> I find this tricky, sql wise
> 
> (...)
> 
>   Maybe you have an idea? if not, I can still persuit this approach if you insist even though I am not so sure it will have much benifit.
> 

I don't insist, if it doesn't make sense to you. TBH the issue is hard to follow given I'm not able to respond in time to to big backlog of things. I trust your judgement :)

How about my first suggestion of `fetchFirstColumn` then?

Reply:
Regarding `fetchFirstColumn`, yes, that is implemented, see 26e6145d9ac897fa2181b69fa42345a83129cdfe

And the foreach loop has been moved from `DoctrineStorage::getUrlsFromUrlLink()` to `EzSystems\EzPlatformRichText\eZ\FieldType\RichText\RichTextStorage\Gateway::getUrlsFromUrlLink()` in c6c0c22428a6948b6134459d98e31de76bc5ee1e and https://github.com/ezsystems/ezplatform-richtext/pull/242/commits/0a3c480163686e3a960ae6337c5807bd963d4a08